Sources/AlamofireXMLRPC/Alamofire+XML.swift

```diff
@@ -11,28 +11,63 @@ import AEXML
 import Alamofire
 
 extension DataRequest {
-    public static func XMLResponseSerializer() -> DataResponseSerializer<AEXMLDocument> {
-        return DataResponseSerializer { request, response, data, error in
-            if let e = error {
-                return .failure(e)
-            }
+    @discardableResult public func responseXML(
+        queue: DispatchQueue = .main,
+        dataPreprocessor: DataPreprocessor = XMLResponseSerializer.defaultDataPreprocessor,
+        emptyResponseCodes: Set<Int> = XMLResponseSerializer.defaultEmptyResponseCodes,
+        emptyRequestMethods: Set<HTTPMethod> = XMLResponseSerializer.defaultEmptyRequestMethods,
+        completionHandler: @escaping (AFDataResponse<AEXMLDocument>) -> Void
+    ) -> Self {
+        response(
+            queue: queue,
+            responseSerializer: XMLResponseSerializer(
+                dataPreprocessor: dataPreprocessor,
+                emptyResponseCodes: emptyResponseCodes,
+                emptyRequestMethods: emptyRequestMethods
+            ),
+            completionHandler: completionHandler)
+    }
+}
 
-            guard let validData = data else {
-                return .failure(AFError.responseValidationFailed(reason: AFError.ResponseValidationFailureReason.dataFileNil))
-            }
+public class XMLResponseSerializer: ResponseSerializer {
+    public let dataPreprocessor: DataPreprocessor
+    public let emptyResponseCodes: Set<Int>
+    public let emptyRequestMethods: Set<HTTPMethod>
+
+    public init(
+        dataPreprocessor: DataPreprocessor = XMLResponseSerializer.defaultDataPreprocessor,
+        emptyResponseCodes: Set<Int> = XMLResponseSerializer.defaultEmptyResponseCodes,
+        emptyRequestMethods: Set<HTTPMethod> = XMLResponseSerializer.defaultEmptyRequestMethods
+    ) {
+        self.dataPreprocessor = dataPreprocessor
+        self.emptyResponseCodes = emptyResponseCodes
+        self.emptyRequestMethods = emptyRequestMethods
+    }
+
+    public func serialize(
+        request: URLRequest?,
+        response: HTTPURLResponse?,
+        data: Data?,
+        error: Error?
+    ) throws -> AEXMLDocument {
+        guard error == nil else { throw error! }
 
-            do {
-                let XML = try AEXMLDocument(xml: validData)
-                return .success(XML)
-            } catch {
-                return .failure(error)
+        guard var data = data, !data.isEmpty else {
+            guard emptyResponseAllowed(forRequest: request, response: response) else {
+                throw AFError.responseSerializationFailed(reason: .inputDataNilOrZeroLength)
             }
+
+            let errorDocument = AEXMLDocument()
+            errorDocument.error = .parsingFailed
```

Comment:
In case of empty response is allowed, I wonder if it is really necessary or make sense to set an error. Maybe you actually need to create a _methodResponse_ node inside that is empty.

Reply:
Good point. I wasn't sure on the pattern for handling errors. I'll update the document/element name to the expected node type.

Reply:
For XMLRPC serialization, I return an empty method response element and for XML serialization I return and empty document. The empty document will still include an error if you request the root element. I also added test to verify this behavior.

Reply:
https://github.com/jonathanfoster/AlamofireXMLRPC/commit/495b21cd2df406f29d03b13cc7ad44e458b222b6